### DISCUSSION TITLE
How install in laravel 9

### DISCUSSION THREAD

**Discussion by kioselsa** [botman/botman#1308] [Q&A]
Helow, my name is Oscar Delgado, First of all I would like to thank you for such a good app. I need  your help, I can't install botman in Laravel 9, mark error:     Problem 1 - botman/botman[2.6.0, ..., 2.6.1] require symfony/http-foundation ^2.8 || ^3.0 || ^4.0 || ^5.0 And I can't find the solution. I would be very grateful if you could help me

**Comment by NowakAdmin** [comment]
Hi, I made it a little dirty way but it works at least.

1st download all repositories that You need 

then I uploaded all in single folder(in my case "botman") to main laravel folder

so in main folder i have:
botman
-botman
-driver-facebook
...

next i added in composer.json for every single repository downloaded this way:
`"repositories": [
        {
            "type": "path",
            "url": "./botman/botman"
        }
],`

Then I replaced some code in every downloaded repository composer.json

for example my botman/botman/composer.json (basicly i moved to php 8.1 and removed not used in laravel 9 repositories that was causing problems
`
    "require": {
        "php": ">=8.1",
        "symfony/http-foundation": "^2.8 || ^3.0 || ^4.0 || ^5.0 || ^6.0",
        "tightenco/collect": "^5.0 || ^6.0 || ^7.0 || ^8.0",
        "opis/closure": "^2.3 || ^3.0",
        "mpociot/pipeline": "^1.0.2",
        "react/socket": "~1.0",
        "spatie/macroable": "*",
        "psr/container": "^1.0 || ^2.0"
    },
    "require-dev": {
        "codeigniter/framework": "~3.0",
        "orchestra/testbench": "^3.4 || ^4.0 || ^5.0 || ^6.0",
        "phpunit/phpunit": "^6.0 || ^7.0 || ^8.0",
        "mockery/mockery": "^1.1",
        "doctrine/cache": "^1.6",
        "symfony/cache": "^3.4.35 || ^4.3.7 || ^5.0",
        "ext-curl": "*",
        "dms/phpunit-arraysubset-asserts": "^0.4.0"
`

and finally:
`composer require botman/botman`
With succeed.

You can also make a fork of this repository and do it in simmilar way, for me better was local file based solution - i have repository sourcecode inside my project and i can easly update if needed.

**Comment by filippotoso** [comment]
### Install Laravel

```
composer create-project laravel/laravel botman-demo
```
###  Install Botman

```
cd botman-demo
composer require botman/botman botman/driver-telegram
```
###  Configure Botman

```
mkdir config\botman
copy vendor\botman\botman\assets\config.php config\botman
copy vendor\botman\driver-telegram\stubs\telegram.php config\botman
```
_Remember to setup the Telegram token in app\config\telegram.php_

###  Add in app\routes\web.php

```php
Route::any('telegram', [App\Http\Controllers\TelegramController::class, 'index'])->name('telegram');
```

###  Ignore CSRF token in app\Http\Middleware\VerifyCsrfToken.php

```php
    protected $except = [
        '/telegram',
    ];
```

### Create file app\Http\Controllers\TelegramController.php

```php
<?php

namespace App\Http\Controllers;

use BotMan\BotMan\BotMan;
use BotMan\BotMan\BotManFactory;
use BotMan\BotMan\Cache\LaravelCache;
use BotMan\BotMan\Drivers\DriverManager;
use BotMan\Drivers\Telegram\TelegramContactDriver;
use BotMan\Drivers\Telegram\TelegramDriver;
use BotMan\Drivers\Telegram\TelegramLocationDriver;
use BotMan\Drivers\Telegram\TelegramPhotoDriver;
use Illuminate\Http\Request;

class TelegramController extends Controller
{
    public function index(Request $request)
    {
        DriverManager::loadDriver(TelegramDriver::class);
        DriverManager::loadDriver(TelegramPhotoDriver::class);
        DriverManager::loadDriver(TelegramLocationDriver::class);
        DriverManager::loadDriver(TelegramContactDriver::class);

        $config = array_merge([
            'config' => config('botman.config')
        ], [
            'web' => config('botman.web', [])
        ]);

        $botman = BotManFactory::create($config, new LaravelCache);

        $botman->hears('Hi', function (BotMan $bot) {
            $bot->reply('Hello!');
        });

        $botman->listen();
    }
}
```

**Reply by Odilbukh** [reply]
How to setup webhook for it ?

**Comment by filippotoso** [comment]
The TelegramController::index() method above is the webhook.
The url is specified in the above routes.php example (/telegram)

**Reply by Odilbukh** [reply]
So to set webhook I need add /telegram to url right? 
Example: https://yourapp.domain/telegram

**Comment by Odilbukh** [comment]
So I have setup webhook! 

When I message to bot for example "Hi". The bot doest not answer. In ngrok inspect the request like "419 unknown status"

![image](https://user-images.githubusercontent.com/22895615/177942761-8e87328e-f305-44d6-bf6e-9a3440ac2bf9.png)

**Comment by filippotoso** [comment]
Ignore CSRF token in app\Http\Middleware\VerifyCsrfToken.php

```php
    protected $except = [
        '/telegram',
    ];

**Reply by Odilbukh** [reply]
Super! it helped. Now bot is works! Thanks!

**Comment by Odilbukh** [comment]
There is error about Loging 

laravel.EMERGENCY: Unable to create configured logger. Using emergency logger. {"exception":"[object] (InvalidArgumentException(code: 0): Log [deprecations] is not defined. at /vendor/laravel/framework/src/Illuminate/Log/LogManager.php:207)

**Reply by filippotoso** [reply]
Has nothing to do with Botman.

**Comment by Odilbukh** [comment]
and also 

laravel.WARNING: Return type of BotMan\BotMan\Messages\Outgoing\Question::jsonSerialize() should either be compatible with JsonSerializable::jsonSerialize(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /vendor/botman/botman/src/Messages/Outgoing/Question.php on line 116

**Comment by Odilbukh** [comment]
laravel.WARNING: strcmp(): Passing null to parameter #2 ($string2) of type string is deprecated in /vendor/botman/driver-telegram/src/TelegramDriver.php on line 163

**Reply by adolfoholzer** [reply]
Hi Odilbukh, did it work? I'm having several issues trying to make it work, the simple instructions are not working for me.

Thanks!

**Reply by Odilbukh** [reply]
> Hi Odilbukh, did it work? I'm having several issues trying to make it work, the simple instructions are not working for me.
> 
> Thanks!

nope
